homeassistant/components/konnected/config_flow.py

```diff
@@ -0,0 +1,732 @@
+        """Handle the initial step."""
+        errors = {}
+        if user_input:
+            try:
+                self.host = user_input[CONF_HOST]
+                self.port = user_input[CONF_PORT]
+
+                # try to obtain the mac address from the device
+                status = await get_status(self.hass, self.host, self.port)
+                self.device_id = status.get("mac").replace(":", "")
+                self.model = status.get("name", KONN_MODEL)
+
+                # finish if we have a cached data set (partial config imported earlier)
+                if self.cached_config:
+                    config = copy.deepcopy(self.cached_config)
+                    try:
+                        config[CONF_HOST] = self.host
+                        config[CONF_PORT] = self.port
+                        return await self.async_create_or_update_entry(
+                            DEVICE_SCHEMA(config)
+                        )
+
+                    except vol.Invalid as err:
+                        _LOGGER.warning(
+                            "Invalid cached config..%s", humanize_error(config, err),
+                        )
+
+                return await self.async_step_io()
+
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+
+        return self.async_show_form(
+            step_id="user", data_schema=vol.Schema(DATA_SCHEMA_MANUAL), errors=errors
+        )
+
+    async def async_step_ssdp(self, discovery_info):
+        """Handle a discovered konnected panel.
+
+        This flow is triggered by the SSDP component. It will check if the
+        device is already configured and attempt to finish the config if not.
+        """
+        from homeassistant.components.ssdp import (
+            ATTR_UPNP_MANUFACTURER,
+            ATTR_UPNP_MODEL_NAME,
+        )
+
+        _LOGGER.debug(discovery_info)
+
+        try:
+            if discovery_info[ATTR_UPNP_MANUFACTURER] != KONN_MANUFACTURER:
+                return self.async_abort(reason="not_konn_panel")
+
+            if not any(
+                name in discovery_info[ATTR_UPNP_MODEL_NAME]
+                for name in KONN_PANEL_MODEL_NAMES
+            ):
+                _LOGGER.warning(
+                    "Discovered unrecognized Konnected device %s",
+                    discovery_info.get(ATTR_UPNP_MODEL_NAME, "Unknown"),
+                )
+                return self.async_abort(reason="not_konn_panel")
+
+            # extract host/port from ssdp_location
+            netloc = urlparse(discovery_info["ssdp_location"]).netloc.split(":")
+            self.host = self.context["host"] = netloc[0]
+            self.port = netloc[1]
+            self.model = discovery_info[ATTR_UPNP_MODEL_NAME]
+
+        except KeyError:
+            _LOGGER.error("Malformed Konnected SSDP info")
+            return self.async_abort(reason="unknown")
+
+        if any(
+            self.host == flow["context"].get("host")
```

Comment:
It's best if we can get the device id already here and try to set a unique_id for the flow. That will automatically abort if it's already set in other flows.

Reply:
Awesome - didn't know about that mechanism. Will update.